src/stream.rs

```diff
@@ -1222,12 +1316,16 @@ impl RangeBuf {
     /// Splits the buffer into two at the given index.
     pub fn split_off(&mut self, at: usize) -> RangeBuf {
         let buf = RangeBuf {
-            data: self.data.split_off(at),
-            pos: 0,
+            data: self.data.clone(),
+            start: self.start + at,
+            pos: cmp::max(self.pos, self.start + at),
+            len: self.len - at,
```

Comment:
do you want a `saturating_sub` here?

Reply:
The idea is that `at` should never be grater than `self.len`, otherwise it's a bug. So crashing here might not be a bad thing I think.

Reply:
Though we could add an explicit check for that.

Reply:
Yeah the std::Vec implementation does this

```
fn assert_failed(at: usize, len: usize) -> ! {
            panic!("`at` split index (is {}) should be <= len (is {})", at, len);
        }

        if at > self.len() {
            assert_failed(at, self.len());
        }
```

Reply:
Fixed.